lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,220 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from django.db.utils import IntegrityError
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from base64 import b64encode
+from collections import namedtuple
+from hashlib import sha256
+import json
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=44, unique=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return "VisibleBlocks object - hash:{}, raw json:'{}'".format(self.hashed, self._blocks_json)
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        _blocks_json = json.dumps([block._asdict() for block in blocks], separators=(',', ':'), sort_keys=True)
+        hashed = b64encode(sha256(_blocks_json).digest())
+        model, _ = cls.objects.get_or_create(hashed=hashed, defaults={'_blocks_json': _blocks_json})
+        return model
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as a list of BlockRecords in the order they were provided.
+        """
+        block_dicts = json.loads(self._blocks_json)
+        return [BlockRecord(data['locator'], data['weight'], data['max_score']) for data in block_dicts]
+
+    @blocks.setter
+    def blocks(self, value):
+        """
+        Not implemented, as VisibleBlocks instances are intended to be write-once and not change after creation.
+        """
+        raise NotImplementedError(
+            "Property 'blocks' cannot be modified on an existing VisibleBlocks model. Create a new instance."
+        )
+
+
+class BlockRecord(namedtuple('BlockRecordBase', ['locator', 'weight', 'max_score'])):
+    """
+    A namedtuple used to serialize information about a block at the time it was used in grade calculation.
+    """
+    pass
+
+
+class PersistentSubsectionGradeManager(models.QuerySet):
+    """
+    A custom QuerySet (to be used as a manager), that handles creating a VisibleBlocks model on creation.
+    """
+    def create(self, **kwargs):
```

Comment:
I'm still not super clear on the case for using `**kwargs` instead of named arguments. Is it just because there are so many, or is there a more technical reason>